### PR TITLE
Vmware ESXi proxy - add hw grains

### PR DESCRIPTION
**Pull request by lomeroe**
Add more detailed hardware/OS grain data to the ESXi minion.

Following grains are set to slightly different values:
- os (set to a more generic "VMware ESXi" to match other OSes (i.e. Windows, RedHat)
- osrelease (set to the ESXi version number instead of proxy)
- osfullname (value of previous "os" grain)
- hwaddr_interfaces (populated with vmnic info)

Following grains are added:
- biosreleasedate
- biosversion
- cpu_model
- domain
- fqdn
- host
- ip4_interfaces
- ip6_interfaces
- ip_interfaces
- manufacturer
- mem_total
- num_cpu_cores
- num_cpu_sockets
- num_cpus
- osbuild
- osmanufacturer
- productname
- serialnumber
- timezone

example grain data before

```
myesx-minion:
    ----------
    apiType:
        HostAgent
    apiVersion:
        5.5
    build:
        2718055
    cpuarch:
        x86_64
    dynamicProperty:
    dynamicType:
        None
    fullName:
        VMware ESXi 5.5.0 build-2718055
    gpus:
    hwaddr_interfaces:
        ----------
    id:
        myesx-minion
    instanceUuid:
        None
    kernel:
        proxy
    kernelrelease:
        proxy
    licenseProductName:
        VMware ESX Server
    licenseProductVersion:
        5.0
    localeBuild:
        000
    localeVersion:
        INTL
    locale_info:
        ----------
    machine_id:
    master:
        salt
    mem_total:
        0
    name:
        VMware ESXi
    nodename:
        my-proxy-minion
    num_gpus:
        0
    os:
        VMware ESXi 5.5.0 build-2718055
    osType:
        vmnix-x86
    os_family:
        proxy
    osarch:
        x86_64
    osrelease:
        proxy
    osrelease_info:
        - proxy
    path:
        /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
    productLineId:
        embeddedEsx
    ps:
        ps -efH
    pythonexecutable:
        /usr/bin/python
    pythonpath:
        - /var/cache/salt/minion/extmods/utils
        - /usr/bin
        - /usr/lib64/python27.zip
        - /usr/lib64/python2.7
        - /usr/lib64/python2.7/plat-linux2
        - /usr/lib64/python2.7/lib-tk
        - /usr/lib64/python2.7/lib-old
        - /usr/lib64/python2.7/lib-dynload
        - /usr/lib64/python2.7/site-packages
        - /usr/lib/python2.7/site-packages
    pythonversion:
        - 2
        - 7
        - 5
        - final
        - 0
    saltpath:
        /usr/lib/python2.7/site-packages/salt
    saltversion:
        2015.8.7
    saltversioninfo:
        - 2015
        - 8
        - 7
        - 0
    vendor:
        VMware, Inc.
    version:
        5.5.0
    virtual:
        VMWare
    zmqversion:
        4.0.5

```

grain data with this commit:

```
myesx-minion:
    ----------
    apiType:
        HostAgent
    apiVersion:
        5.5
    biosreleasedate:
        05/06/2015
    biosversion:
        U17
    build:
        2718055
    cpu_model:
        Intel(R) Xeon(R) CPU E7-8890 v3 @ 2.50GHz
    cpuarch:
        x86_64
    domain:
        my.domain
    dynamicProperty:
    dynamicType:
        None
    fqdn:
        myesx-minion.my.domain
    fullName:
        VMware ESXi 5.5.0 build-2718055
    gpus:
    host:
        myesx-minion
    hwaddr_interfaces:
        ----------
        vmk0:
            28:80:23:xx:xx:xx
        vmnic0:
            5c:b9:01:xx:xx:xx
        vmnic1:
            5c:b9:01:xx:xx:xx
        vmnic2:
            00:11:0a:xx:xx:xx
        vmnic3:
            00:11:0a:xx:xx:xx
        vmnic4:
            28:80:23:xx:xx:xx
        vmnic5:
            28:80:23:xx:xx:xx
        vmnic6:
            28:80:23:xx:xx:xx
        vmnic7:
            28:80:23:xx:xx:xx
        vmnic8:
            00:22:64:xx:xx:xx
        vmnic9:
            00:22:64:xx:xx:xx
    id:
        myesx-minion
    instanceUuid:
        None
    ip4_interfaces:
        ----------
        vmk0:
            - 10.10.10.10
    ip6_interfaces:
        ----------
        vmk0:
    ip_interfaces:
        ----------
        vmk0:
            - 10.10.10.10
    kernel:
        proxy
    kernelrelease:
        proxy
    licenseProductName:
        VMware ESX Server
    licenseProductVersion:
        5.0
    localeBuild:
        000
    localeVersion:
        INTL
    locale_info:
        ----------
    machine_id:
    manufacturer:
        HP
    master:
        salt
    mem_total:
        1048456
    name:
        VMware ESXi
    nodename:
        my-proxy-minion
    num_cpu_cores:
        72
    num_cpu_sockets:
        4
    num_cpus:
        288
    num_gpus:
        0
    os:
        VMware ESXi
    osType:
        vmnix-x86
    os_family:
        proxy
    osarch:
        x86_64
    osbuild:
        2718055
    osfullname:
        VMware ESXi 5.5.0 build-2718055
    osmanufacturer:
        VMware, Inc.
    osrelease:
        5.5.0
    osrelease_info:
        - proxy
    path:
        /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
    productLineId:
        embeddedEsx
    productname:
        ProLiant DL580 Gen9
    ps:
        ps -efH
    pythonexecutable:
        /usr/bin/python
    pythonpath:
        - /var/cache/salt/minion/extmods/utils
        - /usr/bin
        - /usr/lib64/python27.zip
        - /usr/lib64/python2.7
        - /usr/lib64/python2.7/plat-linux2
        - /usr/lib64/python2.7/lib-tk
        - /usr/lib64/python2.7/lib-old
        - /usr/lib64/python2.7/lib-dynload
        - /usr/lib64/python2.7/site-packages
        - /usr/lib/python2.7/site-packages
    pythonversion:
        - 2
        - 7
        - 5
        - final
        - 0
    saltpath:
        /usr/lib/python2.7/site-packages/salt
    saltversion:
        2015.8.7
    saltversioninfo:
        - 2015
        - 8
        - 7
        - 0
    serialnumber:
        XXXXXXXXXX
    timezone:
        UTC
    vendor:
        VMware, Inc.
    version:
        5.5.0
    virtual:
        VMWare
    zmqversion:
        4.0.5
```

### CODE REVIEW

File: salt/utils/vmware.py [hunk 1]
```diff
@@ -307,6 +307,70 @@ def _get_dvs_uplink_portgroup(dvs, portgroup_name):
     return None
 
 
+def get_hardware_grains(service_instance):
+    '''
+    Return hardware info for standard minion grains if the service_instance is a HostAgent type
+
```
Comment (rallytime) [inline salt/utils/vmware.py]: Can you add a ``.. versionadded:: Carbon` tag here?

**Comment by rallytime** [comment]
@lomeroe Excellent! Thanks for improving these grains. :D

I had one _tiny_ comment, and it looks like there is a small [lint error](https://jenkins.saltstack.com/job/salt-pr-lint-n/13664/violations/file/salt/modules/vsphere.py/) as well. Once those are cleaned up we can get this in!

**Comment by cachedout** [comment]
Hi @lomeroe 

What's the reasoning behind the slight changes? They concern me because even small changes in those grains can have wide-ranging impacts, including possibly breaking people's deployments when targets no longer match as expected.

If we're going to change these, I'd like to understand fully the reasoning. Typically, this needs to be a pretty compelling case because the potential for breaking things is so high. Could we have a further discussion about this?

**Comment by lomeroe** [comment]
Sure - it is really only my opinion in how they should be set based on values for those grains on other minions and that they should follow more similar patterns/standards.  I'm not necessarily married to any of these changes, but think they are a good idea.  Also, I think it is better to standardize them now when the ESXi proxy minion is relatively young.

Here's a little more details on my line of thinking:

The "os" grain returns only "Windows", "RedHat", "SUSE", "Ubuntu", etc on minions running those OSes, as opposed to the ESXi proxy minion which returns what other minions would call "osfullname" as "os".  Honestly, after thinking about it "os" would probably even be better set as simply "ESXi".  I listed "osfullname" in the "slight change" category, even though it isn't currently defined by the ESXi proxy minion, only b/c the currently value of "os" in the proxy minion is more akin to the value of "osfullname" on other minions (for example "Red Hat Enterprise Linux Server" or "Microsoft Windows Server 2012 R2 Standard")

Osrelease on all other minions lists the version of the OS (for example "2012Server" on a Windows minion or "7.2" on a RHEL/Cent 7.2 minion, again this seems more logical to return the os release number of the ESXi server as opposed to the string "proxy".  To follow suit "osrelease_info" should probably also be modified (not done in the current PR).

hwaddr_interfaces currently just returns empty (on my hosts, maybe for others it returns data), the change here is to add the virtual hardware interface data, again making it more standard to other minions.

**Comment by cro** [comment]
@lomeroe The challenge here I think is that if we replace all the core grains then we lose the knowledge of the machine running the proxy itself.  What do you think about moving the proxy data to the grains dictionary under a key like 'proxy_grains'?

**Comment by cro** [comment]
I should also mention I think that should become a standard for all proxies.

**Comment by lomeroe** [comment]
@cro I'm not exactly following which core grains you're concerned about losing from the minion running the proxy process, that minion's grains shouldn't bet getting modified.

Here's my setup, with output (using this PR) so you can help me understand what you mean.

I have a docker container running the salt-minon process, this salt-minion process's ID is "esxiproxy"
this container also runs the salt-proxy process for several ESXi hosts.  One with a proxyid of "esxserver1" (i.e. salt-proxy --proxyid=esxserver1)

get the grains of the minion running the proxy process (i.e. the docker container, these grains should not be getting modified by anything here)
```
salt 'esxiproxy' grains.items
esxiproxy:
    ----------
    SSDs:
    cpu_flags:
        - fpu
        - vme
        - de
        - pse
        - tsc
        - msr
        - pae
        - mce
        - cx8
        - apic
        - sep
        - mtrr
        - pge
        - mca
        - cmov
        - pat
        - pse36
        - clflush
        - dts
        - mmx
        - fxsr
        - sse
        - sse2
        - ss
        - ht
        - syscall
        - nx
        - rdtscp
        - lm
        - constant_tsc
        - arch_perfmon
        - pebs
        - bts
        - nopl
        - xtopology
        - tsc_reliable
        - nonstop_tsc
        - aperfmperf
        - pni
        - pclmulqdq
        - ssse3
        - cx16
        - sse4_1
        - sse4_2
        - x2apic
        - popcnt
        - aes
        - hypervisor
        - lahf_lm
        - ida
        - arat
        - epb
        - dtherm
    cpu_model:
        Intel(R) Xeon(R) CPU           X5690  @ 3.47GHz
    cpuarch:
        x86_64
    domain:
        my_domain.com
    fqdn:
        esxiproxy.my_domain.com
    fqdn_ip4:
        - 172.17.0.106
    fqdn_ip6:
        - fe80::42:acff:fe11:6a
    gpus:
        |_
          ----------
          model:
              SVGA II Adapter
          vendor:
              unknown
    host:
        esxiproxy
    hwaddr_interfaces:
        ----------
    id:
        esxiproxy
    init:
        unknown
    ip4_interfaces:
        ----------
    ip6_interfaces:
        ----------
    ip_interfaces:
        ----------
    ipv4:
    ipv6:
    kernel:
        Linux
    kernelrelease:
        3.10.0-229.7.2.el7.x86_64
    locale_info:
        ----------
        defaultencoding:
            None
        defaultlanguage:
            None
        detectedencoding:
            ascii
    localhost:
        esxiproxy
    lsb_distrib_codename:
        CentOS Linux 7 (Core)
    lsb_distrib_id:
        CentOS Linux
    lsb_distrib_release:
        7
    machine_id:
    master:
        salt
    mdadm:
    mem_total:
        15914
    nodename:
        esxiproxy
    num_cpus:
        4
    num_gpus:
        1
    os:
        CentOS
    os_family:
        RedHat
    osarch:
        x86_64
    oscodename:
        CentOS Linux 7 (Core)
    osfinger:
        CentOS Linux-7
    osfullname:
        CentOS Linux
    osmajorrelease:
        7
    osrelease:
        7
    osrelease_info:
        - 7
    path:
        /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
    ps:
        ps -efH
    pythonexecutable:
        /usr/bin/python
    pythonpath:
        - /var/cache/salt/minion/extmods/utils
        - /usr/bin
        - /usr/lib64/python27.zip
        - /usr/lib64/python2.7
        - /usr/lib64/python2.7/plat-linux2
        - /usr/lib64/python2.7/lib-tk
        - /usr/lib64/python2.7/lib-old
        - /usr/lib64/python2.7/lib-dynload
        - /usr/lib64/python2.7/site-packages
        - /usr/lib/python2.7/site-packages
    pythonversion:
        - 2
        - 7
        - 5
        - final
        - 0
    saltpath:
        /usr/lib/python2.7/site-packages/salt
    saltversion:
        2015.8.7
    saltversioninfo:
        - 2015
        - 8
        - 7
        - 0
    server_id:
        207686610
    shell:
        /bin/sh
    systemd:
        ----------
        features:
            +PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ -LZ4 -SECCOMP +BLKID +ELFUTILS +KMOD +IDN
        version:
            219
    virtual:
        VMWare
    zmqversion:
        4.0.5
```

Get the grains of the ESXi host via the salt-proxy process (modified in this PR)
```
salt 'esxserver1' grains.items
esxserver1:
    ----------
    apiType:
        HostAgent
    apiVersion:
        5.5
    biosreleasedate:
        10/01/2013
    biosversion:
        P65
    build:
        2718055
    cpu_model:
        Intel(R) Xeon(R) CPU E7- 4870  @ 2.40GHz
    cpuarch:
        x86_64
    domain:
        my_domain.com
    dynamicProperty:
    dynamicType:
        None
    fqdn:
        esxserver1.my_domain.com
    fullName:
        VMware ESXi 5.5.0 build-2718055
    gpus:
    host:
        esxserver1
    hwaddr_interfaces:
        ----------
        vmk0:
            e8:39:xx:xx:xx:c4
        vmnic0:
            e8:39:xx:xx:xx:c4
        vmnic1:
            e8:39:xx:xx:xx:c5
        vmnic2:
            e8:39:xx:xx:xx:c6
        vmnic3:
            e8:39:xx:xx:xx:c7
        vmnic4:
            2c:27:yy:yy:yy:41
        vmnic5:
            2c:27:yy:yy:yy:a3
    id:
        esxserver1
    instanceUuid:
        None
    ip4_interfaces:
        ----------
        vmk0:
            - 10.10.86.201
    ip6_interfaces:
        ----------
        vmk0:
    ip_interfaces:
        ----------
        vmk0:
            - 10.10.86.201
    kernel:
        proxy
    kernelrelease:
        proxy
    licenseProductName:
        VMware ESX Server
    licenseProductVersion:
        5.0
    localeBuild:
        000
    localeVersion:
        INTL
    locale_info:
        ----------
    machine_id:
    manufacturer:
        HP
    master:
        salt
    mem_total:
        524277
    name:
        VMware ESXi
    nodename:
        esxiproxy
    num_cpu_cores:
        40
    num_cpu_sockets:
        4
    num_cpus:
        160
    num_gpus:
        0
    os:
        VMware ESXi
    osType:
        vmnix-x86
    os_family:
        proxy
    osarch:
        x86_64
    osbuild:
        2718055
    osfullname:
        VMware ESXi 5.5.0 build-2718055
    osmanufacturer:
        VMware, Inc.
    osrelease:
        5.5.0
    osrelease_info:
        - proxy
    path:
        /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
    productLineId:
        embeddedEsx
    productname:
        ProLiant DL580 G7
    ps:
        ps -efH
    pythonexecutable:
        /usr/bin/python
    pythonpath:
        - /var/cache/salt/minion/extmods/utils
        - /usr/bin
        - /usr/lib64/python27.zip
        - /usr/lib64/python2.7
        - /usr/lib64/python2.7/plat-linux2
        - /usr/lib64/python2.7/lib-tk
        - /usr/lib64/python2.7/lib-old
        - /usr/lib64/python2.7/lib-dynload
        - /usr/lib64/python2.7/site-packages
        - /usr/lib/python2.7/site-packages
    pythonversion:
        - 2
        - 7
        - 5
        - final
        - 0
    saltpath:
        /usr/lib/python2.7/site-packages/salt
    saltversion:
        2015.8.7
    saltversioninfo:
        - 2015
        - 8
        - 7
        - 0
    serialnumber:
        USEXXXXXXX
    timezone:
        UTC
    vendor:
        VMware, Inc.
    version:
        5.5.0
    virtual:
        VMWare
    zmqversion:
        4.0.5
```

**Comment by cro** [comment]
> I'm not exactly following which core grains you're concerned about losing from the minion running the proxy process, that minion's grains shouldn't be getting modified.

I didn't word that very well, I'm sorry.

I was foreseeing some Salt users who may not want to run a regular minion on the machine that runs the proxy minion.  If you don't run a regular minion, only a proxy minion, and all the grains get replaced, then you don't have any way to tell anything about the machine that runs the proxy.

For sure I see both sides of the issue, and your approach is very elegant and makes perfect sense in terms of what we want the proxy to be--it should behave in (almost) every way like a regular minion.  It also opens the door to re-use of other salt modules for proxies when those modules depend on core grains.

As I write that last statement I think that's the clincher--in the long run if we do grains like you've set them up here, then it should make the larger Salt codebase a little bit less complicated--fewer situations where we would say "look for this grain here unless you are a proxy, then look here instead".

I'm going to cogitate on this a little more and talk to some other folks.

**Comment by lomeroe** [comment]
Ok, I see what you are saying.  Does it work that way today?  I honestly haven't tried (and actually never thought of doing it :D ).   If I run only the salt-proxy process and no salt-minion process, are the "host" systems grains then aggregated in when querying the salt-proxy's id?  I don't see how the current code base would do that, but I'm probably missing something.

I *personally* wouldn't see that as a function of the salt-proxy process, that process (in my mind) is there to return information/interface with the "dumb" device, as if it was running a full minion.  If I wanted to see grain data from the system running that salt-proxy process, then that system needs a salt-minion process.  It would be nice to tell that it is a proxied minion though, so perhaps all proxies should return a "proxy_minion: True" grain (or something similar)?

**Comment by cro** [comment]
> Ok, I see what you are saying. Does it work that way today? I honestly haven't tried
> (and actually never thought of doing it :D ). If I run only the salt-proxy process and
> no salt-minion process, are the "host" systems grains then aggregated in when 
> querying the salt-proxy's id?

No, there's no communication or check to see if the proxy is running but there is no minion.

> so perhaps all proxies should return a "proxy_minion: True" grain (or something similar)?

We had been relying on the `os` grain for that.  But I think probably a `proxymodule` grain is a better approach, then with one grain you can tell both that it is a proxy and what kind of proxy.

I'm inclined to merge this, I'll write up some documentation for the release notes.

**Pull request merged**